cunumeric/sorting.py

```diff
@@ -0,0 +1,101 @@
+# Copyright 2022 NVIDIA Corporation
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+from cunumeric.config import CuNumericOpCode
+
+from legate.core import types as ty
+
+
+def sort_flattened(output, input, argsort, stable):
+    flattened = input.reshape((input.size,), order="C")
+
+    # run sort flattened -- return 1D solution
+    sort_result = output.runtime.create_empty_thunk(
+        flattened.shape, dtype=output.dtype, inputs=(flattened,)
+    )
+    sorting(sort_result, flattened, argsort, stable=stable)
+    output.base = sort_result.base
+    output.numpy_array = None
+
+
+def sort_swapped(output, input, argsort, sort_axis, stable):
+    assert sort_axis < input.ndim - 1 and sort_axis >= 0
+
+    # swap axes
+    swapped = input.swapaxes(sort_axis, input.ndim - 1)
+
+    swapped_copy = output.runtime.create_empty_thunk(
+        swapped.shape, dtype=input.dtype, inputs=(input, swapped)
+    )
+    swapped_copy.copy(swapped, deep=True)
+
+    # run sort on last axis
+    sort_result = output.runtime.create_empty_thunk(
```

Comment:
Why is this thunk necessary if `swapped_copy` is already a copy we can mutate? Can we do the sorting in place using `swapped_copy`?

Reply:
I tried to keep the logic simple here. The underlying code dose not support input==output at this time. I could change this, but this will still not always be an option (with argsort the input is of different type than the output).

Reply:
The code skips the copy whenever possible.